R/stochastic_process.R

```
@@ -76,84 +74,428 @@ stone_stochastic_process <- function(con, modelling_group, disease,
+  if (is.na(index_from)) {
+    index_from <- 1
+    index_to <- 1
   }
 
-  # Scenario-specific tests:
-  # 1. (touchstone, scenario_description) exists in scenario table
-  # 2. (scenario_description, disease) exists in scenario_description table
-  # 3. (touchstone, modelling_group) exists in responsibility_set table
-  # 4. (responsibility_set, scenario) exist in responsibility, where
-  #    scenario id comes from #1, and responsibility_set id from #3
+  ################################################################
 
-  for (scenario in scenarios) {
-    scenario_id <- DBI::dbGetQuery(con, "
-      SELECT id FROM scenario WHERE touchstone = $1
-         AND scenario_description = $2", list(touchstone, scenario))$id
-    if (length(scenario_id) != 1) {
-      stop(sprintf("scenario %s not found in touchstone %s",
-                   scenario, touchstone))
+  for (i in index_from:index_to) {
+
+    the_file <- read_params$files[scenario_no]
+    the_file <- gsub(":index", i, the_file)
+    the_file <- gsub(":group", touchpoint$modelling_group, the_file)
+    the_file <- gsub(":touchstone", touchpoint$touchstone, the_file)
+    the_file <- gsub(":disease", touchpoint$disease, the_file)
+    the_file <- gsub(":scenario", scenario, the_file)
+
+    if (!file.exists(the_file)) {
+      stop(sprintf("File not found: %s", the_file))
     }
+    message(the_file)
 
-    scenario_descs <- DBI::dbGetQuery(con, "
-      SELECT count(*) FROM scenario_description WHERE id = $1
-         AND disease = $2", list(scenario, disease))$count
-    if (scenario_descs != 1) {
-      stop(sprintf("scenario_description %s not valid for disease %s",
-                   scenario, disease))
+    scenario_data[[i]] <-
+      read_xz_csv(con, the_file, outcomes,
+                  read_params$allow_missing_disease,
+                  read_params$runid_from_file, i,
+                  touchpoint$touchstone, countries)
+  }
```

I'd prefer to keep this file check done for all scenarios before you start doing any work on any of them (the dry_run arg before) - otherwise it could be 8 hours before you realise there's a missing file in the last scenario. So maybe, continuing the pattern for the PR, factor out a function that makes `the_file` with all those gsubs... then copy the index_from:index_to loop (and a surround scenario loop) to the validation function, and fail early if all files don't exist.

---

Good point, thanks. I've moved the file name substitution into validation function and added a check that those files exist. Then passed those files into `process_scenario` so should fail early now if any problems with naming